node/src/tasks/bsp_upload_request.rs

```diff
@@ -0,0 +1,158 @@
+use std::path::Path;
+
+use forest_manager::traits::ForestStorage;
+use sc_tracing::tracing::{error, info, warn};
+use shc_common::types::HasherOutT;
+
+use file_manager::traits::{FileStorage, FileStorageWriteError, FileStorageWriteOutcome};
+use storage_hub_infra::event_bus::EventHandler;
+use tokio::{fs::File, io::AsyncWriteExt};
+
+use crate::services::{
+    file_transfer::{commands::FileTransferServiceInterface, events::RemoteUploadRequest},
+    handler::{StorageHubHandler, StorageHubHandlerConfig},
+};
+
+const LOG_TARGET: &str = "bsp-upload-request-handler";
+
+pub struct BspUploadRequestHandler<SHC: StorageHubHandlerConfig> {
+    storage_hub_handler: StorageHubHandler<SHC>,
+}
+
+impl<SHC: StorageHubHandlerConfig> Clone for BspUploadRequestHandler<SHC> {
+    fn clone(&self) -> BspUploadRequestHandler<SHC> {
+        Self {
+            storage_hub_handler: self.storage_hub_handler.clone(),
+        }
+    }
+}
+
+impl<SHC: StorageHubHandlerConfig> BspUploadRequestHandler<SHC> {
+    pub fn new(storage_hub_handler: StorageHubHandler<SHC>) -> Self {
+        Self {
+            storage_hub_handler,
+        }
+    }
+}
+
+impl<SHC: StorageHubHandlerConfig> EventHandler<RemoteUploadRequest>
+    for BspUploadRequestHandler<SHC>
+where
+    HasherOutT<SHC::TrieLayout>: TryFrom<[u8; 32]>,
+{
+    async fn handle_event(&self, event: RemoteUploadRequest) -> anyhow::Result<()> {
+        if !event.chunk_with_proof.verify() {
+            error!(
+                target: LOG_TARGET,
+                "Received invalid proof for chunk: {} (file: {:?}))", event.chunk_with_proof.proven.key, event.file_key
+            );
+            // TODO: Record this for further reputation actions
+            return Ok(());
+        }
+
+        let file_key: HasherOutT<SHC::TrieLayout> = TryFrom::try_from(*event.file_key.as_ref())
+            .map_err(|_| anyhow::anyhow!("File key and HasherOutT mismatch!"))?;
+
+        let mut write_file_storage = self.storage_hub_handler.file_storage.write().await;
+        let write_chunk_result = write_file_storage.write_chunk(
+            &file_key,
+            &event.chunk_with_proof.proven.key,
+            &event.chunk_with_proof.proven.data,
+        );
+        drop(write_file_storage);
+
+        match write_chunk_result {
+            Ok(outcome) => match outcome {
+                FileStorageWriteOutcome::FileComplete => self.on_file_complete(&file_key).await,
+                FileStorageWriteOutcome::FileIncomplete => {}
+            },
+            Err(error) => match error {
```

Comment:
In any of this cases, shouldn't we also unregister the user and file from the file transfer service?

Reply:
The only ones that are not an internal error are `FileDoesNotExists` or `FileChunkAlreadyExists`.

If the file doesn't exists it means that either:
1. There is a bug and we didn't add any metadata of the file before starting to receive the file
2. The file was deleted by another parallel request

If the file chunk already exists .. should be unregister (and delete file) directly or just update reputation and decide later?